File: ganttproject/src/net/sourceforge/ganttproject/gui/ProjectUIFacadeImpl.java
/*
GanttProject is an opensource project management tool.
Copyright (C) 2005-2011 GanttProject team

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 3
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package net.sourceforge.ganttproject.gui;

import biz.ganttproject.core.option.GPOptionGroup;
import biz.ganttproject.storage.NetworkUnavailableException;
import biz.ganttproject.storage.OnlineDocument;
import biz.ganttproject.storage.OnlineDocumentMode;
import biz.ganttproject.storage.StorageDialogAction;
import biz.ganttproject.storage.VersionMismatchException;
import net.sourceforge.ganttproject.GPLogger;
import net.sourceforge.ganttproject.IGanttProject;
import net.sourceforge.ganttproject.action.CancelAction;
import net.sourceforge.ganttproject.action.GPAction;
import net.sourceforge.ganttproject.action.OkAction;
import net.sourceforge.ganttproject.document.Document;
import net.sourceforge.ganttproject.document.Document.DocumentException;
import net.sourceforge.ganttproject.document.DocumentManager;
import net.sourceforge.ganttproject.document.ProxyDocument;
import net.sourceforge.ganttproject.document.webdav.WebDavStorageImpl;
import net.sourceforge.ganttproject.filter.GanttXMLFileFilter;
import net.sourceforge.ganttproject.gui.projectwizard.NewProjectWizard;
import net.sourceforge.ganttproject.language.GanttLanguage;
import net.sourceforge.ganttproject.undo.GPUndoManager;
import org.eclipse.core.runtime.IStatus;

import javax.swing.*;
import javax.swing.filechooser.FileFilter;
import java.awt.event.ActionEvent;
import java.io.File;
import java.io.IOException;
import java.net.URI;
import java.text.MessageFormat;
import java.util.ArrayList;
import java.util.List;
import java.util.logging.Level;

public class ProjectUIFacadeImpl implements ProjectUIFacade {
  final UIFacade myWorkbenchFacade;
  private final GanttLanguage i18n = GanttLanguage.getInstance();
  private final DocumentManager myDocumentManager;
  private final GPUndoManager myUndoManager;

  private final GPOptionGroup myConverterGroup = new GPOptionGroup("convert", ProjectOpenStrategy.getMilestonesOption());

  public ProjectUIFacadeImpl(UIFacade workbenchFacade, DocumentManager documentManager, GPUndoManager undoManager) {
    myWorkbenchFacade = workbenchFacade;
    myDocumentManager = documentManager;
    myUndoManager = undoManager;
  }

  @Override
  public void saveProject(IGanttProject project) {
    if (project.getDocument() == null) {
      saveProjectAs(project);
      return;
    }
    Document document = project.getDocument();
    saveProjectTryWrite(project, document);
  }

  private boolean saveProjectTryWrite(final IGanttProject project, final Document document) {
    IStatus canWrite = document.canWrite();
    if (!canWrite.isOK()) {
      GPLogger.getLogger(Document.class).log(Level.INFO, canWrite.getMessage(), canWrite.getException());
      String message = formatWriteStatusMessage(document, canWrite);
      List<Action> actions = new ArrayList<Action>();
      actions.add(new GPAction("project.saveas") {
        @Override
        public void actionPerformed(ActionEvent e) {
          saveProjectAs(project);
        }
      });
      if (canWrite.getCode() == Document.ErrorCode.LOST_UPDATE.ordinal()) {
        actions.add(new GPAction("document.overwrite") {
          @Override
          public void actionPerformed(ActionEvent e) {
            saveProjectTryLock(project, document);
          }
        });
      }
      actions.add(CancelAction.EMPTY);
      myWorkbenchFacade.showOptionDialog(JOptionPane.ERROR_MESSAGE, message, actions.toArray(new Action[0]));

      return false;
    }
    return saveProjectTryLock(project, document);
  }

  private boolean saveProjectTryLock(IGanttProject project, Document document) {
    return saveProjectTrySave(project, document);
  }

  private static OnlineDocument asOnlineDocument(Document doc) {
    if (doc instanceof ProxyDocument) {
      ProxyDocument proxyDocument = (ProxyDocument) doc;
      if (proxyDocument.getRealDocument() instanceof OnlineDocument) {
        return (OnlineDocument) proxyDocument.getRealDocument();
      }
    }
    return null;
  }

  private boolean saveProjectTrySave(IGanttProject project, Document document) {
    OnlineDocument onlineDoc = asOnlineDocument(document);
    try {
      saveProject(document);
      afterSaveProject(project);
      if (onlineDoc != null && onlineDoc.getMode() == OnlineDocumentMode.OFFLINE_ONLY) {
        tryConnectOnline(onlineDoc);
      }
      return true;
    } catch (NetworkUnavailableException e) {
      if (onlineDoc != null) {
        myWorkbenchFacade.showOptionDialog(JOptionPane.WARNING_MESSAGE,
            "You seem to be offline. Continue working with offline mirror?",
            new Action[]{new CancelAction(), new OkAction() {
              @Override
              public void actionPerformed(ActionEvent actionEvent) {
                onlineDoc.setMode(OnlineDocumentMode.OFFLINE_ONLY);
              }
            }});
      }
      return false;
    } catch (VersionMismatchException e) {
      if (onlineDoc != null) {
        myWorkbenchFacade.showOptionDialog(JOptionPane.WARNING_MESSAGE,
            "Your document has diverged from the cloud",
            new Action[]{new CancelAction(), new OkAction() {
              @Override
              public void actionPerformed(ActionEvent actionEvent) {

              }
            }});
      }
      return false;
    } catch (Throwable e) {
      myWorkbenchFacade.showErrorDialog(e);
      return false;
    }
  }

  private void tryConnectOnline(OnlineDocument onlineDoc) {
    try {
      onlineDoc.setMode(OnlineDocumentMode.MIRROR);
      myWorkbenchFacade.showOptionDialog(JOptionPane.INFORMATION_MESSAGE, "You are online", new Action[]{new OkAction() {
        @Override
        public void actionPerformed(ActionEvent actionEvent) {

        }
      }});
    } catch (Exception e) {

    }
  }

  private String formatWriteStatusMessage(Document doc, IStatus canWrite) {
    assert canWrite.getCode() >= 0 && canWrite.getCode() < Document.ErrorCode.values().length;
    Document.ErrorCode errorCode = Document.ErrorCode.values()[canWrite.getCode()];
    String key = "document.error.write." + errorCode.name().toLowerCase();
    return MessageFormat.format(i18n.getText(key), doc.getPath(), canWrite.getMessage());
  }

  private void afterSaveProject(IGanttProject project) {
    Document document = project.getDocument();
    myDocumentManager.addToRecentDocuments(document);
    String title = i18n.getText("appliTitle") + " [" + document.getFileName() + "]";
    myWorkbenchFacade.setWorkbenchTitle(title);
    if (document.isLocal()) {
      URI url = document.getURI();
      if (url != null) {
        File file = new File(url);
        myDocumentManager.changeWorkingDirectory(file.getParentFile());
      }
    }
    project.setModified(false);
  }

  private void saveProject(Document document) throws IOException {
    myWorkbenchFacade.setStatusText(GanttLanguage.getInstance().getText("saving") + " " + document.getPath());
    document.write();
  }

  @Override
  public void saveProjectAs(IGanttProject project) {
    new StorageDialogAction(project, myWorkbenchFacade, this, project.getDocumentManager(),
        ((WebDavStorageImpl) project.getDocumentManager().getWebDavStorageUi()).getServersOption()).actionPerformed(null);
  }

  /**
   * Check if the project has been modified, before creating or opening another
   * project
   *
   * @return true when the project is <b>not</b> modified or is allowed to be
   * discarded
   */
  @Override
  public boolean ensureProjectSaved(IGanttProject project) {
    if (project.isModified()) {
      UIFacade.Choice saveChoice = myWorkbenchFacade.showConfirmationDialog(i18n.getText("msg1"),
          i18n.getText("warning"));
      if (UIFacade.Choice.CANCEL == saveChoice) {
        return false;
      }
      if (UIFacade.Choice.YES == saveChoice) {
        try {
          saveProject(project);
          // If all those complex save procedures complete successfully and project gets saved
          // then its modified state becomes false
          // Otherwise it remains true which means we have not saved and can't continue
          return !project.isModified();
        } catch (Exception e) {
          myWorkbenchFacade.showErrorDialog(e);
          return false;
        }
      }
    }
    return true;
  }

  @Override
  public void openProject(final IGanttProject project) throws IOException, DocumentException {
    if (false == ensureProjectSaved(project)) {
      return;
    }
    JFileChooser fc = new JFileChooser(myDocumentManager.getWorkingDirectory());
    FileFilter ganttFilter = new GanttXMLFileFilter();

    // Remove the possibility to use a file filter for all files
    FileFilter[] filefilters = fc.getChoosableFileFilters();
    for (int i = 0; i < filefilters.length; i++) {
      fc.removeChoosableFileFilter(filefilters[i]);
    }
    fc.addChoosableFileFilter(ganttFilter);

    int returnVal = fc.showOpenDialog(myWorkbenchFacade.getMainFrame());
    if (returnVal == JFileChooser.APPROVE_OPTION) {
      Document document = getDocumentManager().getDocument(fc.getSelectedFile().getAbsolutePath());
      openProject(document, project);
    }
  }

  @Override
  public void openProject(final Document document, final IGanttProject project) throws IOException, DocumentException {
    beforeClose();
    project.close();

    try (ProjectOpenStrategy strategy = new ProjectOpenStrategy(project, myWorkbenchFacade)) {
      strategy.openFileAsIs(document)
          .checkLegacyMilestones()
          .checkEarliestStartConstraints()
          .runUiTasks();
    } catch (Exception e) {
      throw new DocumentException("Can't open document " + document, e);
    }
  }

  private void beforeClose() {
    myWorkbenchFacade.setWorkbenchTitle(i18n.getText("appliTitle"));
    getUndoManager().die();
  }

  @Override
  public void createProject(final IGanttProject project) {
    if (false == ensureProjectSaved(project)) {
      return;
    }
    beforeClose();
    project.close();
    myWorkbenchFacade.setStatusText(i18n.getText("project.new.description"));
    showNewProjectWizard(project);
  }

  private void showNewProjectWizard(IGanttProject project) {
    NewProjectWizard wizard = new NewProjectWizard();
    wizard.createNewProject(project, myWorkbenchFacade);
  }

  @Override
  public GPOptionGroup[] getOptionGroups() {
    return new GPOptionGroup[]{myConverterGroup};
  }

  private GPUndoManager getUndoManager() {
    return myUndoManager;
  }

  private DocumentManager getDocumentManager() {
    return myDocumentManager;
  }
}

File: ganttproject/src/main/java/net/sourceforge/ganttproject/gui/ProjectUIFacadeImpl.kt
/*
Copyright 2019 BarD Software s.r.o
Copyright 2005-2018 GanttProject team

This file is part of GanttProject, an opensource project management tool.

GanttProject is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.

GanttProject is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with GanttProject.  If not, see <http://www.gnu.org/licenses/>.
 */
package net.sourceforge.ganttproject.gui

import biz.ganttproject.app.*
import biz.ganttproject.core.calendar.ImportCalendarOption
import biz.ganttproject.core.option.GPOptionGroup
import biz.ganttproject.lib.fx.VBoxBuilder
import biz.ganttproject.storage.*
import biz.ganttproject.storage.cloud.*
import com.google.common.collect.Lists
import com.sandec.mdfx.MDFXNode
import de.jensd.fx.glyphs.fontawesome.FontAwesomeIcon
import de.jensd.fx.glyphs.fontawesome.FontAwesomeIconView
import javafx.application.Platform
import javafx.geometry.Pos
import javafx.scene.layout.BorderPane
import javafx.scene.layout.Pane
import kotlinx.coroutines.*
import kotlinx.coroutines.channels.BroadcastChannel
import kotlinx.coroutines.channels.Channel
import net.sourceforge.ganttproject.GPLogger
import net.sourceforge.ganttproject.IGanttProject
import net.sourceforge.ganttproject.document.Document
import net.sourceforge.ganttproject.document.Document.DocumentException
import net.sourceforge.ganttproject.document.DocumentManager
import net.sourceforge.ganttproject.document.ProxyDocument
import net.sourceforge.ganttproject.document.webdav.WebDavStorageImpl
import net.sourceforge.ganttproject.gui.projectwizard.NewProjectWizard
import net.sourceforge.ganttproject.importer.BufferProject
import net.sourceforge.ganttproject.importer.asImportBufferProjectApi
import net.sourceforge.ganttproject.importer.importBufferProject
import net.sourceforge.ganttproject.language.GanttLanguage
import net.sourceforge.ganttproject.resource.HumanResourceMerger
import net.sourceforge.ganttproject.resource.HumanResourceMerger.MergeResourcesOption.BY_ID
import net.sourceforge.ganttproject.task.event.*
import net.sourceforge.ganttproject.task.export
import net.sourceforge.ganttproject.task.importFromDatabase
import net.sourceforge.ganttproject.undo.GPUndoManager
import java.io.File
import java.io.IOException
import java.util.concurrent.Executors
import java.util.logging.Level
import javax.swing.SwingUtilities


@ExperimentalCoroutinesApi
class ProjectUIFacadeImpl(
    private val myWorkbenchFacade: UIFacade,
    private val documentManager: DocumentManager,
    private val undoManager: GPUndoManager) : ProjectUIFacade {
  private val i18n = GanttLanguage.getInstance()

  private val myConverterGroup = GPOptionGroup("convert", ProjectOpenStrategy.milestonesOption)
  private var isSaving = false

  override fun saveProject(project: IGanttProject, onFinish: Channel<Boolean>?) {
    if (isSaving) {
      GPLogger.logToLogger("We're saving the project now. This save request was rejected")
    }
    isSaving = true
    try {
      val broadcastWaitScope = CoroutineScope(Executors.newFixedThreadPool(2).asCoroutineDispatcher())
      val broadcastChannel = BroadcastChannel<Boolean>(1)
      broadcastChannel.openSubscription().let { channel ->
        broadcastWaitScope.launch {
          if (channel.receive()) {
            afterSaveProject(project)
            channel.cancel()
          }
        }
      }
      onFinish?.let {
        broadcastChannel.openSubscription().let { channel -> broadcastWaitScope.launch {
          it.send(channel.receive())
          channel.cancel()
        }}
      }
      ProjectSaveFlow(project = project, onFinish = broadcastChannel,
        signin = this::signin,
        error = this::onError,
        saveAs = { saveProjectAs(project) }
      ).run()
    } finally {
      isSaving = false
    }
  }

  fun onError(ex: Exception) {
    dialog {
      it.addStyleSheet("/biz/ganttproject/app/dialogs.css")
      it.addStyleClass("dialog-alert")
      it.setHeader(
        VBoxBuilder("header").apply {
          addTitle(RootLocalizer.create("error.channel.itemTitle")).also { hbox ->
            hbox.alignment = Pos.CENTER_LEFT
            hbox.isFillHeight = true
          }
        }.vbox
      )
      it.setContent(
        //createAlertBody(ex.message ?: ""),
        MDFXNode(ex.message ?: "").also { it.styleClass.add("content-pane") }
      )
      it.removeButtonBar()
    }
  }

  override fun saveProjectAs(project: IGanttProject) {
    StorageDialogAction(
      project, this, project.documentManager,
      (project.documentManager.webDavStorageUi as WebDavStorageImpl).serversOption,
      StorageDialogBuilder.Mode.SAVE,
      "project.save"
    ).doRun()
  }

  enum class CantWriteChoice {MAKE_COPY, CANCEL, RETRY}

  private fun signin(onAuth: ()->Unit) {
    dialog { controller ->
      val wrapper = BorderPane()
      controller.addStyleClass("dlg-lock", "dlg-cloud-file-options")
      controller.addStyleSheet(
        "/biz/ganttproject/storage/cloud/GPCloudStorage.css",
        "/biz/ganttproject/storage/StorageDialog.css"
      )
      wrapper.center = Pane().also {
        it.prefHeight = 400.0
        it.prefWidth = 400.0
      }
      controller.setContent(wrapper)
      GPCloudUiFlowBuilder().apply {
        flowPageChanger = createFlowPageChanger(wrapper, controller)
        mainPage = object : EmptyFlowPage() {
          override var active: Boolean
            get() = super.active
            set(value) {
              if (value) {
                controller.hide()
                onAuth()
              }
            }
        }
        build().start()
      }
    }
  }
//  private fun formatWriteStatusMessage(doc: Document, canWrite: IStatus): String {
//    assert(canWrite.code >= 0 && canWrite.code < Document.ErrorCode.values().size)
//    return RootLocalizer.formatText(
//        key = "document.error.write.${Document.ErrorCode.values()[canWrite.code].name.toLowerCase()}",
//        doc.fileName, canWrite.message)
//  }

  private fun afterSaveProject(project: IGanttProject) {
    val document = project.document
    documentManager.addToRecentDocuments(document)
    val title = i18n.getText("appliTitle") + " [" + document.fileName + "]"
    myWorkbenchFacade.setWorkbenchTitle(title)
    if (document.isLocal) {
      val url = document.uri
      if (url != null) {
        val file = File(url)
        documentManager.changeWorkingDirectory(file.parentFile)
      }
    }
    project.isModified = false
  }

  /**
   * Check if the project has been modified, before creating or opening another
   * project
   *
   * @return true when the project is **not** modified or is allowed to be
   * discarded
   */
  override fun ensureProjectSaved(project: IGanttProject): Barrier<Boolean> =
    if (project.isModified) {
      val saveChoice = myWorkbenchFacade.showConfirmationDialog(i18n.getText("msg1"),
          i18n.getText("warning"))
      when (saveChoice) {
        UIFacade.Choice.CANCEL, null -> ResolvedBarrier(false)
        UIFacade.Choice.NO -> ResolvedBarrier(true)
        UIFacade.Choice.YES, UIFacade.Choice.OK -> {
          SimpleBarrier<Boolean>().also { barrier ->
            val onFinish = Channel<Boolean>()
            CoroutineScope(Executors.newSingleThreadExecutor().asCoroutineDispatcher()).launch {
              try {
                if (onFinish.receive()) {
                  barrier.resolve(true)
                }
              } catch (e: Exception) {
                myWorkbenchFacade.showErrorDialog(e)
              }
            }
            saveProject(project, onFinish)
          }
        }
      }
    } else {
      ResolvedBarrier(true)
    }


  @Throws(IOException::class, DocumentException::class)
  override fun openProject(document: Document, project: IGanttProject, onFinish: Channel<Boolean>?,
                           authenticationFlow: AuthenticationFlow?) {
    try {
      ProjectOpenStrategy(
        project = project,
        uiFacade = myWorkbenchFacade,
        signin = authenticationFlow ?: this::signin,
      ).use { strategy ->
        DOCUMENT_LOGGER.debug(">>> openProject({})", document.uri)
        // Run coroutine which fetches document and wait until it sends the result to the channel.
        val docChannel = Channel<Document>()
        CoroutineScope(Executors.newSingleThreadExecutor().asCoroutineDispatcher()).launch {
          try {
            DOCUMENT_LOGGER.debug("... waiting for the document")
            docChannel.receive().also { doc ->
              DOCUMENT_LOGGER.debug("... document is ready")
              // If document is obtained, we need to run further steps.
              // Because of historical reasons they run in Swing thread (they may modify the state of Swing components)
              SwingUtilities.invokeLater {
                try {
                  beforeClose()
                  project.close()
                  strategy.openFileAsIs(doc)
                    .checkLegacyMilestones()
                    .checkEarliestStartConstraints()
                    .runUiTasks()
                  document.asOnlineDocument()?.let {
                    if (it is GPCloudDocument) {
                      it.colloboqueClient = ColloboqueClient(project.projectDatabase, undoManager)
                      project.projectDatabase.addExternalUpdatesListener {
                        Platform.runLater {
                          val emptyTaskManager = project.taskManager.emptyClone()
                          emptyTaskManager.importFromDatabase(project.projectDatabase.readAllTasks(), project.taskManager.taskHierarchy.export())
                          val bufferProject = BufferProject(
                            emptyTaskManager,
                            project.projectDatabase,
                            project.roleManager,
                            project.activeCalendar,
                            myWorkbenchFacade
                          )
                          val mergeOption = HumanResourceMerger.MergeResourcesOption()
                          val importCalendarOption = ImportCalendarOption()
                          mergeOption.loadPersistentValue(BY_ID)
                          importCalendarOption.loadPersistentValue(ImportCalendarOption.Values.REPLACE.name)
                          importBufferProject(
                            project,
                            bufferProject,
                            myWorkbenchFacade.asImportBufferProjectApi(),
                            mergeOption,
                            importCalendarOption,
                            closeCurrentProject = true
                          )
                        }
                      }
                      it.onboard(documentManager, webSocket)
                    }
                  }
                  runBlocking { onFinish?.send(true) }
                } catch (ex: DocumentException) {
                  onFinish?.close(ex) ?: DOCUMENT_ERROR_LOGGER.error("", ex)
                }
              }
            }
          } catch (ex: Exception) {
            when (ex) {
              // If channel was closed with a cause and it was because of HTTP 403, we show UI for sign-in
              is DocumentException -> {
                onFinish?.close(ex) ?: DOCUMENT_ERROR_LOGGER.error("", ex)
              }
              else -> {
                onFinish?.close(ex) ?: DOCUMENT_ERROR_LOGGER.error("Can't open document $document", ex)
              }
            }
          }
          finally {
            DOCUMENT_LOGGER.debug("<<< openProject()")
          }
        }
        strategy.open(document, docChannel)
      }
    } catch (e: Exception) {
      throw DocumentException("Can't open document $document", e)
    }
  }

  private fun beforeClose() {
    myWorkbenchFacade.setWorkbenchTitle(i18n.getText("appliTitle"))
    undoManager.die()
  }

  override fun createProject(project: IGanttProject) {
    ensureProjectSaved(project).await { result ->
      if (result) {
        beforeClose()
        project.close()
        myWorkbenchFacade.setStatusText(i18n.getText("project.new.description"))
        showNewProjectWizard(project)
      }
    }
  }

  private fun showNewProjectWizard(project: IGanttProject) {
    val wizard = NewProjectWizard()
    wizard.createNewProject(project, myWorkbenchFacade)
  }

  override fun getOptionGroups(): Array<GPOptionGroup> {
    return arrayOf(myConverterGroup)
  }
}

@ExperimentalCoroutinesApi
class ProjectSaveFlow(
    private val project: IGanttProject,
    private val onFinish: BroadcastChannel<Boolean>,
    private val signin: (()->Unit) -> Unit,
    private val error: (Exception) -> Unit,
    private val saveAs: () -> Unit) {

  private val doneScope = CoroutineScope(Executors.newSingleThreadExecutor().asCoroutineDispatcher())
  private fun done(success: Boolean) {
    doneScope.launch {
      onFinish.send(success)
      onFinish.close()
    }
  }

  fun run() {
    try {
      project.document?.let {
        if (it.asLocalDocument()?.canRead() == false) {
          saveProjectAs(project)
        } else {
          if (it is ProxyDocument) {
            it.createContents()
          }
          saveProjectTryWrite(project, it)
        }
      } ?: run {
        saveProjectAs(project)
      }
    } catch (ex: Exception) {
      error(ex)
      done(success = false)
    }
  }


  private fun saveProjectTryWrite(project: IGanttProject, document: Document) {
    val canWrite = document.canWrite()
    if (!canWrite.isOK) {
      GPLogger.getLogger(Document::class.java).log(Level.INFO, canWrite.message, canWrite.exception)
      OptionPaneBuilder<ProjectUIFacadeImpl.CantWriteChoice>().also {
        it.i18n = RootLocalizer.createWithRootKey(
          rootKey = "document.error.write.cantWrite",
          baseLocalizer = RootLocalizer
        )
        it.styleClass = "dlg-lock"
        it.styleSheets.add("/biz/ganttproject/storage/cloud/GPCloudStorage.css")
        it.styleSheets.add("/biz/ganttproject/storage/StorageDialog.css")
        it.titleString.update(document.fileName)
        it.titleHelpString?.update(canWrite.message)
        it.graphic = FontAwesomeIconView(FontAwesomeIcon.LOCK, "64").also { icon ->
          icon.styleClass.add("img")
        }
        it.elements = listOf(
          OptionElementData("document.option.makeCopy", ProjectUIFacadeImpl.CantWriteChoice.MAKE_COPY, true),
          OptionElementData("cancel", ProjectUIFacadeImpl.CantWriteChoice.CANCEL, false),
          OptionElementData("generic.retry", ProjectUIFacadeImpl.CantWriteChoice.RETRY, false),
        )
        it.showDialog { choice ->
          SwingUtilities.invokeLater {
            when (choice) {
              ProjectUIFacadeImpl.CantWriteChoice.MAKE_COPY -> {
                saveProjectAs(project)
              }
              ProjectUIFacadeImpl.CantWriteChoice.RETRY -> {
                saveProjectTryWrite(project, document)
              }
              else -> {
                done(success = false)
              }
            }
          }
        }
      }
    } else {
      saveProjectTryLock(project, document)
    }
  }

  private fun saveProjectTryLock(project: IGanttProject, document: Document) {
    saveProjectTrySave(project, document)
  }

  enum class VersionMismatchChoice { OVERWRITE, MAKE_COPY }

  private fun saveProjectTrySave(project: IGanttProject, document: Document) {
    try {
      saveProject(document)
    } catch (e: VersionMismatchException) {
      done(success = false)
      val onlineDoc = document.asOnlineDocument()
      if (onlineDoc != null) {
        OptionPaneBuilder<VersionMismatchChoice>().also {
          it.i18n = RootLocalizer.createWithRootKey(rootKey = "cloud.versionMismatch", baseLocalizer = RootLocalizer)
          it.styleClass = "dlg-lock"
          it.styleSheets.add("/biz/ganttproject/storage/cloud/GPCloudStorage.css")
          it.styleSheets.add("/biz/ganttproject/storage/StorageDialog.css")
          it.graphic = FontAwesomeIconView(FontAwesomeIcon.CODE_FORK, "64").also {icon ->
            icon.styleClass.add("img")
          }
          it.elements = Lists.newArrayList(
            OptionElementData("document.option.makeCopy", VersionMismatchChoice.MAKE_COPY, true)
          ).also { list ->
            if (e.canOverwrite) {
              list.add(OptionElementData("option.overwrite", VersionMismatchChoice.OVERWRITE, false))
            }
          }
          it.showDialog { choice ->
            SwingUtilities.invokeLater {
              when (choice) {
                VersionMismatchChoice.OVERWRITE -> {
                  onlineDoc.write(force = true)
                }
                VersionMismatchChoice.MAKE_COPY -> {
                  saveProjectAs(project)
                }
              }
            }
          }

        }
      }
    } catch (e: ForbiddenException) {
      signin {
        saveProjectTrySave(project, document)
      }
    } catch (e: PaymentRequiredException) {
      done(success = false)
      error(e)
    }
  }

  @Throws(IOException::class)
  private fun saveProject(document: Document) {
    //myWorkbenchFacade.setStatusText(GanttLanguage.getInstance().getText("saving") + " " + document.path)
    document.write()
    done(success = true)
  }

  private fun saveProjectAs(project: IGanttProject) {
    done(success = false)
    saveAs()
  }
}

private val DOCUMENT_LOGGER = GPLogger.create("Document.Info")
private val DOCUMENT_ERROR_LOGGER = GPLogger.create("Document.Error")
